src/NuGet.Clients/NuGet.SolutionRestoreManager/SolutionRestoreWorker.cs

```diff
@@ -217,27 +244,36 @@ private void SolutionEvents_AfterClosing()
             // Initialize if not already done.
             await InitializeAsync();
 
-            if (_solutionManager.Value.IsSolutionFullyLoaded)
-            {
-                // start background runner if not yet started
-                // ignore the value
-                var ignore = _backgroundJobRunner.Value;
-            }
-
             var pendingRestore = _pendingRestore;
 
             // on-board request onto pending restore operation
             _pendingRequests.Value.TryAdd(request);
 
-            // Await completion of the requested restore operation.
-            // The caller will be unblocked immediately upon
-            // cancellation request via provided token.
-            // Method returns false in case of cancellation is requested.
-            var cancellationTcs = new TaskCompletionSource<bool>();
-            using (token.Register(() => cancellationTcs.SetResult(false)))
-            using (_joinableCollection.Join())
+            Logger.LogInformation($"Scheduled restore request {pendingRestore}.");
+
+            try
             {
-                return await await Task.WhenAny(pendingRestore.Task, cancellationTcs.Task);
+                using (_joinableCollection.Join())
+                {
+                    // Await completion of the requested restore operation or
+                    // preliminary termination of the job runner.
+                    // The caller will be unblocked immediately upon
+                    // cancellation request via provided token.
+                    return await await Task
+                        .WhenAny(
+                            pendingRestore.Task,
+                            _backgroundJobRunner.GetValueAsync())
+                        .WithCancellation(token);
+                }
+            }
+            catch(Exception e)
+            {
+                Logger.LogError(e.ToString());
```

Comment:
Do you want to log an error when the token is canceled? Or should you catch OperationCanceledException and swallow it somehow before this catch block so it's not reported as an error?